### PR TITLE
api: Split `TrainSchedule` input and output

### LINKED ISSUE #3257: Split input and ouput in `models/train_schedule.py`
A dummy split as it is commented in https://github.com/DGEXSolutions/osrd/blob/dev/api/osrd_infra/models/train_schedule.py#L14 is a good first step to achieve https://github.com/DGEXSolutions/osrd/issues/3187

### PR DESCRIPTION
**Pull request by bloussou**
This PR is splitting `TrainScheduleModel` in `TrainScheduleOutput` and `TrainScheduleInput` with a 1-1 relationship.

image

close #3257

### CODE REVIEW

File: api/osrd_infra/models/train_schedule.py [hunk 1]
```diff
@@ -22,9 +33,6 @@ class TrainScheduleModel(models.Model):
     )
     speed_limit_composition = models.CharField(max_length=128, null=True)
     options = models.JSONField(null=True, blank=True, validators=[PydanticValidator(TrainScheduleOptions)])
-
-    # Simulation outputs
-    mrsp = models.JSONField(validators=[PydanticValidator(MRSP)])
-    base_simulation = models.JSONField()
-    eco_simulation = models.JSONField(null=True)
-    modes_and_profiles = models.JSONField(default=list)
+    trainscheduleoutput = models.OneToOneField(
```
Comment (Castavo) [inline api/osrd_infra/models/train_schedule.py]: Why not 
```suggestion
    train_schedule_output = models.OneToOneField(
```
?

File: api/osrd_infra/models/train_schedule.py [hunk 2]
```diff
@@ -2,16 +2,27 @@
...
 
-class TrainScheduleModel(models.Model):
+class TrainScheduleOutput(models.Model):
+    mrsp = models.JSONField(validators=[PydanticValidator(MRSP)])
+    base_simulation = models.JSONField()
+    eco_simulation = models.JSONField(null=True)
+    modes_and_profiles = models.JSONField(default=list)
+
+
+class TrainScheduleInput(models.Model):
```
Comment (Castavo) [inline api/osrd_infra/models/train_schedule.py]: Why keep `TrainSchedule` in the name of both class ? Why not `TrainSchedule` and `SimulationResult` for example ?

File: api/osrd_infra/models/train_schedule.py (hunk 1 above)
Comment (bloussou) [reply api/osrd_infra/models/train_schedule.py]: Indeed looks like in our current code we are considering python convetion > sql convention, I will change that

File: api/osrd_infra/views/train_schedule/train_schedule.py [hunk 3]
```diff
@@ -129,8 +130,14 @@ def standalone_simulation(self, request):
         response_payload = run_simulation(request_payload)
 
         # Process simulation response
-        process_simulation_response(train_schedules[0].timetable.infra, train_schedules, response_payload)
+        simulation_outputs = process_simulation_response(
+            train_schedules[0].timetable.infra, train_schedules, response_payload
+        )
+
+        with transaction.atomic():
```
Comment (Castavo) [inline api/osrd_infra/views/train_schedule/train_schedule.py]: Nice

**Review by Castavo — DISMISSED** [review]
Nice

File: api/osrd_infra/models/train_schedule.py [hunk 4]
```diff
@@ -27,9 +25,6 @@ class TrainScheduleModel(models.Model):
...
     options = models.JSONField(null=True, blank=True, validators=[PydanticValidator(TrainScheduleOptions)])
-
-    # Simulation outputs
-    mrsp = models.JSONField(validators=[PydanticValidator(MRSP)])
-    base_simulation = models.JSONField()
-    eco_simulation = models.JSONField(null=True)
-    modes_and_profiles = models.JSONField(default=list)
+    simulation_output = models.OneToOneField(
+        SimulationOutput, on_delete=models.CASCADE, related_name="train_schedule", null=True
+    )
```
Comment (eckter) [inline api/osrd_infra/models/train_schedule.py]: The new class has all the inputs directly at the root level, and the outputs in a new field at the same level?

This causes some weird semantics in other place (see other comments), and it isn't the best at splitting the inputs/outputs: the output is now "part" of the inputs.

I think I would have preferred if we had "inputs" and "outputs" classes, with this TrainSchedule combining one instance of each.

File: api/osrd_infra/serializers.py [hunk 5]
```diff
@@ -170,26 +170,17 @@ class Meta:
...
-            model = TrainScheduleModel
+            model = TrainSchedule
             exclude = [
                 "timetable",
                 "path",
-                "mrsp",
-                "base_simulation",
-                "eco_simulation",
-                "modes_and_profiles",
             ]
```
Comment (eckter) [inline api/osrd_infra/serializers.py]: I think this is supposed to exclude the outputs from the serialization when creating the payload to the back-end. We probably need to add something in there. But I'm not 100% sure and can't really check.

File: api/osrd_infra/tests/test_stdcm.py [hunk 6]
```diff
@@ -78,23 +79,26 @@ def setUp(self):
...
+        )
+        self.schedule_input = TrainSchedule.objects.create(
+            timetable=self.timetable,
             departure_time=0,
             initial_speed=0,
-            mrsp=[],
             rolling_stock=self.rolling_stock,
-            eco_simulation={},
+            simulation_output=self.schedule_output,
         )
```
Comment (eckter) [inline api/osrd_infra/tests/test_stdcm.py]: `schedule_input` contains the `schedule_output`? Either the naming is off, or something is semantically wrong here.

File: api/osrd_infra/views/train_schedule/standalone_simulation.py [hunk 7]
```diff
@@ -83,7 +83,40 @@ def run_simulation(request_payload):
...
+        modes_and_profiles=modes_and_profiles or [],
+    )
+
+    # Skip if no eco simulation is available
+    if not eco_simulation:
+        return simulation_output
+
+    # Update stops (adding id and name when available)
+    for stop, stop_update in zip(eco_simulation["stops"], stops_updates):
+        stop.update(stop_update)
```
Comment (eckter) [inline api/osrd_infra/views/train_schedule/standalone_simulation.py]: It seems like this block shouldn't be here (duplicated and weirdly placed), if it should it needs an extra comment

File: tests/tests/stdcm/empty_timetable.py [hunk 8]
```diff
@@ -14,7 +14,7 @@ def run(*args, **kwargs):
     infra_id = kwargs["all_scenarios"]["small_infra"].infra
     project = create_project(base_url)
     op_study = create_op_study(base_url, project)
-    scenario, timetable = create_scenario(base_url, infra_id, project, op_study)
+    _, timetable = create_scenario(base_url, infra_id, project, op_study)
```
Comment (eckter) [inline tests/tests/stdcm/empty_timetable.py:17]: It seems unrelated but why not

File: api/osrd_infra/models/train_schedule.py (hunk 4 above)
Comment (bloussou) [reply api/osrd_infra/models/train_schedule.py]: not sure we need to have input and output classes, I think `TrainSchedule` and `SimulationOutput` are enough for now. I agree with you on the point that the relation was in an incorrect order before sorry for that I have changed it.

**Review by eckter — DISMISSED** [review]

**Review by eckter — APPROVED** [review]

**Review by multun — APPROVED** [review]

**Review by flomonster — APPROVED** [review]

**Pull request merged**